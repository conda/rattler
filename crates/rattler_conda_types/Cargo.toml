[package]
name = "rattler_conda_types"
version = "0.27.2"
edition.workspace = true
authors = ["Bas Zalmstra <zalmstra.bas@gmail.com>"]
description = "Rust data types for common types used within the Conda ecosystem"
categories.workspace = true
homepage.workspace = true
repository.workspace = true
license.workspace = true
readme.workspace = true

[dependencies]
chrono = { workspace = true }
file_url = { path = "../file_url", version = "0.1.4" }
fxhash = { workspace = true }
glob = { workspace = true }
hex = { workspace = true }
itertools = { workspace = true }
lazy-regex = { workspace = true }
nom = { workspace = true }
purl = { workspace = true, features = ["serde"] }
rattler_digest = { path = "../rattler_digest", version = "1.0.1", default-features = false, features = ["serde"] }
rattler_macros = { path = "../rattler_macros", version = "1.0.1", default-features = false }
regex = { workspace = true }
simd-json = { workspace = true , features = ["serde_impl"]}
serde = { workspace = true, features = ["derive", "rc"] }
serde_json = { workspace = true }
serde_repr = { workspace = true }
serde_with = { workspace = true, features = ["indexmap_2"] }
serde-untagged = { workspace = true }
serde_yaml = { workspace = true }
smallvec = { workspace = true, features = ["serde", "const_new", "const_generics", "union"] }
strum = { workspace = true, features = ["derive"] }
thiserror = { workspace = true }
tracing = { workspace = true }
typed-path = { workspace = true }
url = { workspace = true, features = ["serde"] }
indexmap = { workspace = true }
rattler_redaction = { version = "0.1.1", path = "../rattler_redaction" }
dirs = { workspace = true }

[dev-dependencies]
rand = { workspace = true }
insta = { workspace = true, features = ["yaml", "redactions", "toml", "glob", "filters"] }
rattler_package_streaming = { path = "../rattler_package_streaming", default-features = false, features = ["rustls-tls"] }
tempfile = { workspace = true }
rstest = { workspace = true }
assert_matches = { workspace = true }
hex-literal = { workspace = true }
criterion = { workspace = true, features = ["html_reports"] }
pathdiff = { workspace = true }
dunce = { workspace = true }
tools = { path = "../tools" }

[[bench]]
name = "parse"
harness = false

[[bench]]
name = "prefix_record_from_path"
harness = false
