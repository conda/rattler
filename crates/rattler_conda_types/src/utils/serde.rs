use chrono::{DateTime, Utc};
use fxhash::FxHashMap;
use serde::de::Error as _;
use serde::ser::Error;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::{DeserializeAs, SerializeAs};
use std::collections::BTreeMap;
use std::marker::PhantomData;
use std::path::{Path, PathBuf};
use url::Url;

/// A helper struct that serializes Paths in a normalized way.
/// - Backslashes are replaced with forward-slashes.
pub(crate) struct NormalizedPath;

impl<P: AsRef<Path>> SerializeAs<P> for NormalizedPath {
    fn serialize_as<S>(source: &P, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match source.as_ref().to_str() {
            Some(s) => s.replace('\\', "/").serialize(serializer),
            None => Err(S::Error::custom("path contains invalid UTF-8 characters")),
        }
    }
}

impl<'de> DeserializeAs<'de, PathBuf> for NormalizedPath {
    fn deserialize_as<D>(deserializer: D) -> Result<PathBuf, D::Error>
    where
        D: Deserializer<'de>,
    {
        PathBuf::deserialize(deserializer)
    }
}

/// Deserialize a sequence into `Vec<T>` but filter `None` values.
pub(crate) struct VecSkipNone<T>(PhantomData<T>);

impl<'de, T, I> DeserializeAs<'de, Vec<T>> for VecSkipNone<I>
where
    I: DeserializeAs<'de, Vec<Option<T>>>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<T>, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ok(I::deserialize_as(deserializer)?
            .into_iter()
            .flatten()
            .collect())
    }
}

/// A helper type parser that tries to parse Urls that could be malformed.
pub(crate) struct LossyUrl;

impl<'de> DeserializeAs<'de, Option<Url>> for LossyUrl {
    fn deserialize_as<D>(deserializer: D) -> Result<Option<Url>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let str = match Option::<String>::deserialize(deserializer)? {
            Some(url) => url,
            None => return Ok(None),
        };
        let url = match Url::parse(&str) {
            Ok(url) => url,
            Err(e) => {
                tracing::warn!("unable to parse '{}' as an URL: {e}. Skipping...", str);
                return Ok(None);
            }
        };
        Ok(Some(url))
    }
}

/// A helper type that parses a string either as a string or a vector of strings.
pub(crate) struct MultiLineString;

impl<'de> DeserializeAs<'de, String> for MultiLineString {
    fn deserialize_as<D>(deserializer: D) -> Result<String, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum Inner {
            String(String),
            Multi(Vec<String>),
        }

        Ok(match Inner::deserialize(deserializer)? {
            Inner::String(s) => s,
            Inner::Multi(s) => s.join("\n"),
        })
    }
}

pub(crate) struct Timestamp;

impl<'de> DeserializeAs<'de, chrono::DateTime<chrono::Utc>> for Timestamp {
    fn deserialize_as<D>(deserializer: D) -> Result<chrono::DateTime<chrono::Utc>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let timestamp = i64::deserialize(deserializer)?;

        // Convert from milliseconds to seconds
        let microseconds = if timestamp > 253_402_300_799 {
            timestamp * 1_000
        } else {
            timestamp * 1_000_000
        };

        // Convert the timestamp to a UTC timestamp
        chrono::DateTime::from_timestamp_micros(microseconds)
            .ok_or_else(|| D::Error::custom("got invalid timestamp, timestamp out of range"))
    }
}

impl SerializeAs<chrono::DateTime<chrono::Utc>> for Timestamp {
    fn serialize_as<S>(source: &DateTime<Utc>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        // Convert the date to a timestamp
        let timestamp: i64 = source.timestamp_millis();

        // Determine the precision of the timestamp.
        let timestamp = if timestamp % 1000 == 0 {
            timestamp / 1000
        } else {
            timestamp
        };

        // Serialize the timestamp
        timestamp.serialize(serializer)
    }
}

/// A helper struct to deserialize types from a string without checking the string.
pub struct DeserializeFromStrUnchecked;

/// A helper function used to sort map alphabetically when serializing.
pub(crate) fn sort_map_alphabetically<T: Serialize, S: serde::Serializer>(
    value: &FxHashMap<String, T>,
    serializer: S,
) -> Result<S::Ok, S::Error> {
    value
        .iter()
        .collect::<BTreeMap<_, _>>()
        .serialize(serializer)
}
