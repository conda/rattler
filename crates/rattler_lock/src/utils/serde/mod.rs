mod match_spec_map_or_vec;
mod ordered;
mod pep440_map_or_vec;
mod timestamp;
pub(crate) mod url_or_path;
pub(crate) use match_spec_map_or_vec::MatchSpecMapOrVec;
pub(crate) use ordered::Ordered;
pub(crate) use pep440_map_or_vec::Pep440MapOrVec;
pub(crate) use timestamp::Timestamp;
