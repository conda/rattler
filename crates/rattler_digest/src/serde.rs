//! Provides custom serialization/deserialization functions for [`Output`] of a [`Digest`]
//! Use the struct [`SerializableHash`] to easily serialize the digest.
//!
//! # Example:
//!
//! ```
//!
//! use rattler_digest::serde::SerializableHash;
//!
//! let hash = SerializableHash::<sha2::Sha256>(
//! rattler_digest::parse_digest_from_hex::<sha2::Sha256>("fe51de6107f9edc7aa4f786a70f4a883943bc9d39b3bb7307c04c41410990726").unwrap());
//! let str = serde_json::to_string(&hash).unwrap();
//! let hash: SerializableHash<sha2::Sha256> = serde_json::from_str(&str).unwrap();
//!
//! ```
//!
use digest::{Digest, Output};
use serde::de::Error;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::{DeserializeAs, SerializeAs};
use std::borrow::Cow;
use std::fmt::LowerHex;
use std::ops::Deref;

/// Deserialize the [`Output`] of a [`Digest`].
///
/// If the deserializer is human-readable, it will parse the digest from a hex
/// string. Otherwise, it will deserialize raw bytes.
pub fn deserialize<'de, D, Dig: Digest>(deserializer: D) -> Result<Output<Dig>, D::Error>
where
    D: Deserializer<'de>,
{
    if deserializer.is_human_readable() {
        let str = Cow::<'de, str>::deserialize(deserializer)?;
        super::parse_digest_from_hex::<Dig>(str.as_ref())
            .ok_or_else(|| Error::custom("failed to parse digest"))
    } else {
        Output::<Dig>::deserialize(deserializer)
    }
}

/// Serializes the [`Output`] of a [`Digest`].
///
/// If the serializer is human-readable, it will write the digest as a hex
/// string. Otherwise, it will deserialize raw bytes.
pub fn serialize<'a, S: Serializer, Dig: Digest>(
    digest: &'a Output<Dig>,
    s: S,
) -> Result<S::Ok, S::Error>
where
    &'a Output<Dig>: LowerHex,
{
    if s.is_human_readable() {
        format!("{digest:x}").serialize(s)
    } else {
        digest.serialize(s)
    }
}

/// Wrapper type for easily serializing a Hash
#[derive(Debug, Clone)]
pub struct SerializableHash<T: Digest>(pub Output<T>);

impl<T: Digest> Serialize for SerializableHash<T>
where
    Output<T>: LowerHex,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize::<S, T>(&self.0, serializer)
    }
}

impl<'de, T: Digest + Default> Deserialize<'de> for SerializableHash<T> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let hash_output: Output<T> = deserialize::<D, T>(deserializer)?;
        Ok(SerializableHash(hash_output))
    }
}

impl<T: Digest> From<Output<T>> for SerializableHash<T> {
    fn from(output: Output<T>) -> Self {
        SerializableHash(output)
    }
}

impl<T: Digest> From<SerializableHash<T>> for Output<T> {
    fn from(s: SerializableHash<T>) -> Self {
        s.0
    }
}

impl<T: Digest> Deref for SerializableHash<T> {
    type Target = Output<T>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T: Digest> SerializeAs<Output<T>> for SerializableHash<T>
where
    for<'a> &'a Output<T>: LowerHex,
{
    fn serialize_as<S>(source: &Output<T>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize::<S, T>(source, serializer)
    }
}

impl<'de, T: Digest + Default> DeserializeAs<'de, Output<T>> for SerializableHash<T> {
    fn deserialize_as<D>(deserializer: D) -> Result<Output<T>, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserialize::<D, T>(deserializer)
    }
}

#[cfg(test)]
mod test {
    use crate::serde::SerializableHash;

    #[test]
    pub fn test_serializable_hash() {
        let hash = SerializableHash::<sha2::Sha256>(
            crate::parse_digest_from_hex::<sha2::Sha256>(
                "fe51de6107f9edc7aa4f786a70f4a883943bc9d39b3bb7307c04c41410990726",
            )
            .unwrap(),
        );
        let str = serde_json::to_string(&hash).unwrap();
        let _hash: SerializableHash<sha2::Sha256> = serde_json::from_str(&str).unwrap();
    }
}
