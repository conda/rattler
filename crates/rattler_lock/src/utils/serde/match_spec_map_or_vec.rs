use fxhash::FxBuildHasher;
use indexmap::IndexMap;
use rattler_conda_types::{
    match_spec::package_name_matcher::PackageNameMatcher, MatchSpec, NamelessMatchSpec, PackageName,
};
use serde::{Deserialize, Deserializer};
use serde_with::{serde_as, DeserializeAs, DisplayFromStr};

pub(crate) struct MatchSpecMapOrVec;

impl<'de> DeserializeAs<'de, Vec<String>> for MatchSpecMapOrVec {
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<String>, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[serde_as]
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum MapOrVec {
            Vec(Vec<String>),
            Map(
                #[serde_as(as = "IndexMap<_, DisplayFromStr, FxBuildHasher>")]
                IndexMap<PackageName, NamelessMatchSpec, FxBuildHasher>,
            ),
        }

        Ok(match MapOrVec::deserialize(deserializer)? {
            MapOrVec::Vec(v) => v,
            MapOrVec::Map(m) => m
                .into_iter()
                .map(|(name, spec)| {
                    MatchSpec::from_nameless(spec, Some(PackageNameMatcher::Exact(name)))
                        .to_string()
                })
                .collect(),
        })
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use serde::Deserialize;
    use serde_with::serde_as;

    #[test]
    fn test_parse_dependencies_as_map_or_vec() {
        #[serde_as]
        #[derive(Deserialize, Eq, PartialEq, Debug)]
        struct Data {
            #[serde_as(deserialize_as = "MatchSpecMapOrVec")]
            dependencies: Vec<String>,
        }

        let data_with_map: Data = serde_yaml::from_str("dependencies:\n  foo: \">3.12\"").unwrap();
        let data_with_vec: Data = serde_yaml::from_str("dependencies:\n- \"foo >3.12\"").unwrap();
        assert_eq!(data_with_map, data_with_vec);
    }
}
