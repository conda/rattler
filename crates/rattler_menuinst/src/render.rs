//! This should take a `serde_json` file, render it with all variables and then load it as a `MenuInst` struct

use rattler_conda_types::Platform;
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, path::Path};

#[derive(Debug, Clone, Serialize, Deserialize, Hash, PartialEq, Eq)]
#[serde(transparent)]
pub struct PlaceholderString(pub String);

impl PlaceholderString {
    pub fn resolve(&self, placeholder: impl AsRef<HashMap<String, String>>) -> String {
        replace_placeholders(self.0.clone(), placeholder.as_ref())
    }
}

pub struct BaseMenuItemPlaceholders {
    placeholders: HashMap<String, String>,
}

impl BaseMenuItemPlaceholders {
    pub fn new(base_prefix: &Path, prefix: &Path, platform: Platform) -> Self {
        let dist_name = |p: &Path| {
            p.file_name()
                .map_or_else(|| "empty".to_string(), |s| s.to_string_lossy().to_string())
        };

        let (python, base_python) = if platform.is_windows() {
            (prefix.join("python.exe"), base_prefix.join("python.exe"))
        } else {
            (prefix.join("bin/python"), base_prefix.join("bin/python"))
        };

        let mut vars = HashMap::from([
            ("BASE_PREFIX", base_prefix.to_path_buf()),
            ("PREFIX", prefix.to_path_buf()),
            ("PYTHON", python),
            ("BASE_PYTHON", base_python),
            ("MENU_DIR", prefix.join("Menu")),
            ("HOME", dirs::home_dir().unwrap_or_default()),
        ]);

        if platform.is_windows() {
            vars.insert("BIN_DIR", prefix.join("Library/bin"));
            vars.insert("SCRIPTS_DIR", prefix.join("Scripts"));
            vars.insert("BASE_PYTHONW", base_prefix.join("pythonw.exe"));
            vars.insert("PYTHONW", prefix.join("pythonw.exe"));
        } else {
            vars.insert("BIN_DIR", prefix.join("bin"));
        }

        if platform.is_osx() {
            vars.insert("PYTHONAPP", prefix.join("python.app/Contents/MacOS/python"));
        }

        let mut vars: HashMap<String, String> = vars
            .into_iter()
            .map(|(k, v)| (k.to_string(), v.to_string_lossy().to_string()))
            .collect();

        let icon_ext = if platform.is_windows() {
            "ico"
        } else if platform.is_osx() {
            "icns"
        } else {
            "png"
        };
        vars.insert("ICON_EXT".to_string(), icon_ext.to_string());

        vars.insert("DISTRIBUTION_NAME".to_string(), dist_name(prefix));
        vars.insert("ENV_NAME".to_string(), dist_name(prefix));

        BaseMenuItemPlaceholders { placeholders: vars }
    }

    /// Insert the menu item location into the placeholders
    ///
    /// - On Linux, this is the path to the `.desktop` file
    /// - On Windows, this is the path to the start menu `.lnk` file
    /// - On macOS, this is the path to the `.app` bundle
    pub fn refine(&self, menu_item_location: &Path) -> MenuItemPlaceholders {
        let mut vars = self.placeholders.clone();
        vars.insert(
            "MENU_ITEM_LOCATION".to_string(),
            menu_item_location.to_string_lossy().to_string(),
        );
        MenuItemPlaceholders { placeholders: vars }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct MenuItemPlaceholders {
    placeholders: HashMap<String, String>,
}

impl AsRef<HashMap<String, String>> for MenuItemPlaceholders {
    fn as_ref(&self) -> &HashMap<String, String> {
        &self.placeholders
    }
}

impl AsRef<HashMap<String, String>> for BaseMenuItemPlaceholders {
    fn as_ref(&self) -> &HashMap<String, String> {
        &self.placeholders
    }
}

/// Replace placeholders in a string with values from a hashmap
/// This only replaces placeholders in the form of {{ key }} (note: while this looks like a Jinja template, it is not).
fn replace_placeholders(mut text: String, replacements: &HashMap<String, String>) -> String {
    for (key, value) in replacements {
        let placeholder = format!("{{{{ {key} }}}}");
        text = text.replace(&placeholder, value);
    }
    text
}
