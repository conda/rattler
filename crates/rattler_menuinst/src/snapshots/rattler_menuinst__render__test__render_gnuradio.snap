---
source: crates/rattler_menuinst/src/render.rs
expression: schema
---
MenuInstSchema {
    id: "https://schemas.conda.io/menuinst-1.schema.json",
    schema: "https://json-schema.org/draft-07/schema",
    menu_name: "MNE-Python (__PKG_VERSION__)",
    menu_items: [
        MenuItem {
            name: "Spyder (MNE)",
            description: "The Spyder development environment",
            command: [
                "will be overridden in platforms section",
            ],
            icon: Some(
                "/home/prefix/menu/mne_spyder.png",
            ),
            precommand: None,
            precreate: None,
            working_dir: None,
            activate: Some(
                true,
            ),
            terminal: Some(
                false,
            ),
            platforms: Platforms {
                linux: Some(
                    Linux {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "spyder",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        categories: Some(
                            [
                                "Science",
                            ],
                        ),
                        dbus_activatable: None,
                        generic_name: None,
                        hidden: None,
                        implements: None,
                        keywords: None,
                        mime_type: None,
                        no_display: None,
                        not_show_in: None,
                        only_show_in: None,
                        prefers_non_default_gpu: None,
                        startup_notify: None,
                        startup_wm_class: None,
                        try_exec: None,
                        glob_patterns: None,
                    },
                ),
                osx: Some(
                    MacOS {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "spyder",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        cf_bundle_display_name: Some(
                            "Spyder (MNE-Python __PKG_VERSION__)",
                        ),
                        cf_bundle_identifier: None,
                        cf_bundle_name: Some(
                            "Spyder (MNE-Python __PKG_VERSION__)",
                        ),
                        cf_bundle_spoken_name: None,
                        cf_bundle_version: Some(
                            "__PKG_VERSION__",
                        ),
                        cf_bundle_url_types: None,
                        cf_bundle_document_types: None,
                        ls_application_category_type: None,
                        ls_background_only: None,
                        ls_environment: None,
                        ls_minimum_system_version: None,
                        ls_multiple_instances_prohibited: None,
                        ls_requires_native_execution: None,
                        ns_supports_automatic_graphics_switching: None,
                        ut_exported_type_declarations: None,
                        ut_imported_type_declarations: None,
                        entitlements: None,
                        link_in_bundle: None,
                        event_handler: None,
                    },
                ),
                win: Some(
                    Windows {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "{{ PYTHONW }}",
                                "{{ SCRIPTS_DIR }}\\spyder-script.py",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        desktop: Some(
                            false,
                        ),
                        quicklaunch: None,
                        terminal_profile: None,
                        url_protocols: None,
                        file_extensions: None,
                        app_user_model_id: None,
                    },
                ),
            },
        },
        MenuItem {
            name: "System Info (MNE)",
            description: "Information on the MNE-Python runtime environment",
            command: [
                "/home/prefix/bin/python",
                "/home/prefix/menu/mne_sys_info.py",
            ],
            icon: Some(
                "/home/prefix/menu/mne_info.png",
            ),
            precommand: None,
            precreate: None,
            working_dir: None,
            activate: Some(
                true,
            ),
            terminal: Some(
                true,
            ),
            platforms: Platforms {
                linux: Some(
                    Linux {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        categories: Some(
                            [
                                "Science",
                            ],
                        ),
                        dbus_activatable: None,
                        generic_name: None,
                        hidden: None,
                        implements: None,
                        keywords: None,
                        mime_type: None,
                        no_display: None,
                        not_show_in: None,
                        only_show_in: None,
                        prefers_non_default_gpu: None,
                        startup_notify: None,
                        startup_wm_class: None,
                        try_exec: None,
                        glob_patterns: None,
                    },
                ),
                osx: Some(
                    MacOS {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        cf_bundle_display_name: Some(
                            "System Information (MNE-Python __PKG_VERSION__)",
                        ),
                        cf_bundle_identifier: None,
                        cf_bundle_name: Some(
                            "System Information (MNE-Python __PKG_VERSION__)",
                        ),
                        cf_bundle_spoken_name: None,
                        cf_bundle_version: Some(
                            "__PKG_VERSION__",
                        ),
                        cf_bundle_url_types: None,
                        cf_bundle_document_types: None,
                        ls_application_category_type: None,
                        ls_background_only: None,
                        ls_environment: None,
                        ls_minimum_system_version: None,
                        ls_multiple_instances_prohibited: None,
                        ls_requires_native_execution: None,
                        ns_supports_automatic_graphics_switching: None,
                        ut_exported_type_declarations: None,
                        ut_imported_type_declarations: None,
                        entitlements: None,
                        link_in_bundle: None,
                        event_handler: None,
                    },
                ),
                win: Some(
                    Windows {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        desktop: Some(
                            false,
                        ),
                        quicklaunch: None,
                        terminal_profile: None,
                        url_protocols: None,
                        file_extensions: None,
                        app_user_model_id: None,
                    },
                ),
            },
        },
        MenuItem {
            name: "Prompt (MNE)",
            description: "MNE-Python console prompt",
            command: [
                "will be overridden in platforms section",
            ],
            icon: Some(
                "/home/prefix/menu/mne_console.png",
            ),
            precommand: None,
            precreate: None,
            working_dir: None,
            activate: Some(
                true,
            ),
            terminal: Some(
                true,
            ),
            platforms: Platforms {
                linux: Some(
                    Linux {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "exec",
                                "bash",
                                "--init-file",
                                "/home/prefix/menu/mne_open_prompt.sh",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        categories: Some(
                            [
                                "Science",
                            ],
                        ),
                        dbus_activatable: None,
                        generic_name: None,
                        hidden: None,
                        implements: None,
                        keywords: None,
                        mime_type: None,
                        no_display: None,
                        not_show_in: None,
                        only_show_in: None,
                        prefers_non_default_gpu: None,
                        startup_notify: None,
                        startup_wm_class: None,
                        try_exec: None,
                        glob_patterns: None,
                    },
                ),
                osx: Some(
                    MacOS {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "osascript",
                                "/home/prefix/menu/mne_open_prompt.applescript",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        cf_bundle_display_name: Some(
                            "Prompt (MNE-Python __PKG_VERSION__)",
                        ),
                        cf_bundle_identifier: None,
                        cf_bundle_name: Some(
                            "Prompt (MNE-Python __PKG_VERSION__)",
                        ),
                        cf_bundle_spoken_name: None,
                        cf_bundle_version: Some(
                            "__PKG_VERSION__",
                        ),
                        cf_bundle_url_types: None,
                        cf_bundle_document_types: None,
                        ls_application_category_type: None,
                        ls_background_only: None,
                        ls_environment: None,
                        ls_minimum_system_version: None,
                        ls_multiple_instances_prohibited: None,
                        ls_requires_native_execution: None,
                        ns_supports_automatic_graphics_switching: None,
                        ut_exported_type_declarations: None,
                        ut_imported_type_declarations: None,
                        entitlements: None,
                        link_in_bundle: None,
                        event_handler: None,
                    },
                ),
                win: Some(
                    Windows {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "%SystemRoot%\\system32\\cmd.exe",
                                "/K",
                                "/home/prefix/menu\\mne_open_prompt.bat",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        desktop: Some(
                            false,
                        ),
                        quicklaunch: None,
                        terminal_profile: None,
                        url_protocols: None,
                        file_extensions: None,
                        app_user_model_id: None,
                    },
                ),
            },
        },
        MenuItem {
            name: "Tutorials (MNE)",
            description: "MNE-Python online tutorials",
            command: [
                "will be overridden in platforms section",
            ],
            icon: Some(
                "/home/prefix/menu/mne_web.png",
            ),
            precommand: None,
            precreate: None,
            working_dir: None,
            activate: Some(
                false,
            ),
            terminal: Some(
                false,
            ),
            platforms: Platforms {
                linux: Some(
                    Linux {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "xdg-open",
                                "https://mne.tools/stable/auto_tutorials/",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        categories: Some(
                            [
                                "Science",
                            ],
                        ),
                        dbus_activatable: None,
                        generic_name: None,
                        hidden: None,
                        implements: None,
                        keywords: None,
                        mime_type: None,
                        no_display: None,
                        not_show_in: None,
                        only_show_in: None,
                        prefers_non_default_gpu: None,
                        startup_notify: None,
                        startup_wm_class: None,
                        try_exec: None,
                        glob_patterns: None,
                    },
                ),
                osx: Some(
                    MacOS {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "open",
                                "https://mne.tools/stable/auto_tutorials/",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        cf_bundle_display_name: Some(
                            "Tutorials (MNE-Python __PKG_VERSION__)",
                        ),
                        cf_bundle_identifier: None,
                        cf_bundle_name: Some(
                            "Tutorials (MNE-Python __PKG_VERSION__)",
                        ),
                        cf_bundle_spoken_name: None,
                        cf_bundle_version: Some(
                            "__PKG_VERSION__",
                        ),
                        cf_bundle_url_types: None,
                        cf_bundle_document_types: None,
                        ls_application_category_type: None,
                        ls_background_only: None,
                        ls_environment: None,
                        ls_minimum_system_version: None,
                        ls_multiple_instances_prohibited: None,
                        ls_requires_native_execution: None,
                        ns_supports_automatic_graphics_switching: None,
                        ut_exported_type_declarations: None,
                        ut_imported_type_declarations: None,
                        entitlements: None,
                        link_in_bundle: None,
                        event_handler: None,
                    },
                ),
                win: Some(
                    Windows {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "%SystemRoot%\\system32\\WindowsPowerShell\\v1.0\\powershell.exe",
                                "\"start https://mne.tools/stable/auto_tutorials -WindowStyle hidden\"",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        desktop: Some(
                            false,
                        ),
                        quicklaunch: None,
                        terminal_profile: None,
                        url_protocols: None,
                        file_extensions: None,
                        app_user_model_id: None,
                    },
                ),
            },
        },
        MenuItem {
            name: "User Forum (MNE)",
            description: "MNE-Python forum for discussions, problem solving, and information exchange",
            command: [
                "will be overridden in platforms section",
            ],
            icon: Some(
                "/home/prefix/menu/mne_forum.png",
            ),
            precommand: None,
            precreate: None,
            working_dir: None,
            activate: Some(
                false,
            ),
            terminal: Some(
                false,
            ),
            platforms: Platforms {
                linux: Some(
                    Linux {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "xdg-open",
                                "https://mne.discourse.group",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        categories: Some(
                            [
                                "Science",
                            ],
                        ),
                        dbus_activatable: None,
                        generic_name: None,
                        hidden: None,
                        implements: None,
                        keywords: None,
                        mime_type: None,
                        no_display: None,
                        not_show_in: None,
                        only_show_in: None,
                        prefers_non_default_gpu: None,
                        startup_notify: None,
                        startup_wm_class: None,
                        try_exec: None,
                        glob_patterns: None,
                    },
                ),
                osx: Some(
                    MacOS {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "open",
                                "https://mne.discourse.group",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        cf_bundle_display_name: Some(
                            "Forum (MNE-Python __PKG_VERSION__)",
                        ),
                        cf_bundle_identifier: None,
                        cf_bundle_name: Some(
                            "Forum (MNE-Python __PKG_VERSION__)",
                        ),
                        cf_bundle_spoken_name: None,
                        cf_bundle_version: Some(
                            "__PKG_VERSION__",
                        ),
                        cf_bundle_url_types: None,
                        cf_bundle_document_types: None,
                        ls_application_category_type: None,
                        ls_background_only: None,
                        ls_environment: None,
                        ls_minimum_system_version: None,
                        ls_multiple_instances_prohibited: None,
                        ls_requires_native_execution: None,
                        ns_supports_automatic_graphics_switching: None,
                        ut_exported_type_declarations: None,
                        ut_imported_type_declarations: None,
                        entitlements: None,
                        link_in_bundle: None,
                        event_handler: None,
                    },
                ),
                win: Some(
                    Windows {
                        base: BasePlatformSpecific {
                            name: "",
                            description: "",
                            icon: None,
                            command: [
                                "%SystemRoot%\\system32\\WindowsPowerShell\\v1.0\\powershell.exe",
                                "\"start https://mne.discourse.group -WindowStyle hidden\"",
                            ],
                            working_dir: None,
                            precommand: None,
                            precreate: None,
                            activate: None,
                            terminal: None,
                        },
                        desktop: Some(
                            false,
                        ),
                        quicklaunch: None,
                        terminal_profile: None,
                        url_protocols: None,
                        file_extensions: None,
                        app_user_model_id: None,
                    },
                ),
            },
        },
    ],
}
